src/methods/get-web-flow-authorization-url.ts

```diff
@@ -17,27 +17,11 @@ export function getWebFlowAuthorizationUrlWithState(
   state: State,
   options: any
 ): any {
-  let allowSignup: boolean;
-  if (options.allowSignup === undefined && state.allowSignup === undefined) {
-    allowSignup = true;
-  } else if (
-    options.allowSignup === undefined &&
-    state.allowSignup !== undefined
-  ) {
-    allowSignup = state.allowSignup;
-  } else if (
-    state.allowSignup === undefined &&
-    options.allowSignup !== undefined
-  ) {
-    allowSignup = options.allowSignup;
-  } else {
-    allowSignup = options.allowSignup || state.allowSignup;
-  }
   const optionsWithDefaults = {
     clientId: state.clientId,
     request: state.octokit.request,
     ...options,
-    allowSignup,
+    allowSignup: state.allowSignup ?? options.allowSignup,
     redirectUrl: options.redirectUrl || state.redirectUrl,
     scopes: options.scopes || state.defaultScopes,
```

Comment:
Thanks for simplifying the logic!

Now my only concern is, well, the `redirectUrl` and `scopes` lines.

Overriding `state.redirectUrl` or `state.defaultScopes` may be a security hole: the `oauth/login` url is public, attackers may require wider scopes or specify different redirect url, which may (or may not) surprises clients of `oauth-app.js`.

Reply:
We need to decide for both the `redirectUrl` and `scope` if they should use simple `OR`s or nullish coalescence.

Moreover, putting absolute preference of these values on server side would mean that clients won't be able to customize these values if the server is not under their control. I can't think of a business use case but it might certainly be a genuine requirement.

Also, if allowing these values from the client can be a security hole, then they should be made mandatory on the server OauthApp constructor which in turn would make the ability to pass these values from the client side useless.

Let me know if my thought process is going in the correct direction or not.

Reply:
Case 1: Alice relies on `octokit-app`, she wants absolute control over `scopes` and `redirectUrl`, so she specifies them in the constructor, and she don’t want someone in the dark to gain wider `scopes` or redirect to some other URLs via her API endpoint.

Case 2: Alice wants the `oauth/login` API to be as flexible as possible, so she left `scopes` and `redirectUrl` unspecified from the constructor, and allow anyone to set them directly when calling the API.

Do you think these two cases are both valid use cases?

If these two cases covers 100%, we can just rely on `state.redirectUrl ?? options.redirectUrl`. If not, there should be a feature request.

PS: for `redirectUrl` and `scopes`, valid constructor values are URL strings and arrays, respectively. Both are truthy, so we could use `??` for better consistency and all test cases should pass.

Reply:
> Regarding redirect URLs, no one should be able to redirect to an URL that is not listed in the app settings page on GitHub.

That is correct.

>  Same for scopes, scopes wider than those given in the app settings shouldn't be possible anyway.

There is a difference between OAuth Apps and GitHub Apps. OAuth Apps don't set any scops (see [create new OAuth App form](https://github.com/settings/applications/new)), setting default scopes only implemented in the constructor. But I don't see a risk here, the user has to grant access explicitly for each scope, there is no way to circumvent that.

GitHub Apps don't have a concept of scopes, it's entirely ignored. GitHub Apps have permissions for repositories, organizations, and users. The latter is what sets the permission you can do with an OAuth token created through a GitHub App (GitHub refers to these as "user-to-server tokens"). The permissions cannot be altered on creation, they can only be scoped down with a secondary request.

So in this PR, we can use nullish coalescence for all three options for consistency, it's only something we can do now since we no longer support Node 12. Everything else I'd consider out of scope, we can discuss it separately

Reply:
Hi, I was able to get back to this PR once again and have pushed a new commit which does coalescence like @gr2m proposes. I am making this PR ready to release for now.